### ISSUE TITLE
Proper way to load cache on start

### ISSUE THREAD

**Issue by bojanv55** [ben-manes/caffeine#417]
I implemented CacheLoader interface to be used with LoadingCache like so:

```
@Bean
	public LoadingCache<Integer, someDto> someCache(){
		LoadingCache<Integer, LdaSport> cache = Caffeine.newBuilder()
				.refreshAfterWrite(30, TimeUnit.MINUTES)
				.build(someCacheLoader());
		cache.getAll(Collections.singletonList(-1));
		return cache;
	}
```

What is proper way to initialize cache at start time? I want to start application only if I have full cache populated - and then during the runtime of an app I can refresh every 30 minutes (but if that refresh fails - nothing happens - old values should not be invalidated and remain inside cache). Is there some retry policy if cache refresh fails (I fetch cache values from REST call).

I found only this artifical `cache.getAll(Collections.singletonList(-1));` method call.

**Comment by ben-manes** [comment]
This is a case where you shouldn't use a caching library. You can a volatile, immutable map which you reload on via a scheduled thread. From Spring tutorials it seems that it would look something likem


```java
@Component
public class DtoCache {
  private volatile Map<Integer, Dto> cache;

  @Scheduled(fixedDelay = 3600000)
  public void refresh() {
    var data = // reload
    // Set atomically replace w/o gaps during reload 
    cache = data;
  }
}
```

**Comment by bojanv55** [comment]
Let me try to describe use-case and why I think I need a cache. 

Application is constantly running and connected to AMQP broker in order to process messages. On start of the application I need all the cache values present (or at least when first message arrives - but I do not want to realize that I cannot do cache-load and process message only at that point at time - better at start of an app). If application loads cache on start, and then after that it cannot reload cache values (because network problem etc.) - still OK - it will manage to process some of the messages and some of the messages will be discarded. Now, it also happens that new messages that arrive from AMQP have new info that is still not in cache.- at this point in time I can go to REST and load this data and cache it for further use. And, of course, I need to reload data periodically.

Do you still think that I would be better with a map?

One more question - what happens when REST services for cache re-load are not accessible during run-time? Does it have any back-off strategy, or it waits for next `.refreshAfterWrite(30, TimeUnit.MINUTES)`?

**Comment by ben-manes** [comment]
TLDR - I wouldn't use Caffeine here because it doesn't offer anything helpful. Instead I'd use a `Map` by either swapping or updating its contents, using a periodic thread to do that. 
---

`refreshAfterWrite` does not proactively reload the data, as it is often used with a bounded like expiration. It is to hide latencies if a popular item expires, such as a configuration value, so that user-facing requests don't observe this penalty. If the entry is not popular then it is allowed to expire and be evicted, so it only reacts when the entry is accessed. This means it will return stale entries after the refresh interval until the reload succeeds, including on the initial call that triggers the reload. When combined with expiration, where the data is considered unusable, this behavior makes perfect sense. It is less desirable when used independently.

If a reload fails then the entry is not modified (value or metadata) and the exception is logged. There is no backoff strategy, etc. as that becomes more prescriptive, and we try to keep this to feel more like a fancy `ConcurrentHashMap`.

Since Caffeine is key-value oriented, a case like yours could lead to many per-key requests when entries become eligible. Since you want all of the data without a bound, you likely can make an optimized batch call instead like you showed. As this has to be periodic then that scheduling would be in your code. This leaves very little for the cache to assist in, as you might use `computeIfAbsent` on a miss to avoid stampedes, and so you can use a map instead of this library.

**Comment by bojanv55** [comment]
OK, I will follow your advice then...
